---
title: "Cancer Genomics Lecture"
output: github_document
---

Load bio3d and read in our fasta files for WT & Mutant p53.
```{r}
library(bio3d)
seqs <- read.fasta("lecture18_sequences.fa")
seqs
```

Make sure WT & Mutant p53 sequences are appropriately aligned.
```{r}
alignment <- seqaln(seqs)
alignment
```


```{r}
str_alignment <- str(alignment)
str_alignment
```



Calculate positional identity scores then identify mutated residues.
```{r}
ide <- conserv(seqs$ali, method = "identity")
#ide
mutant.sites <- which(ide < 1)

gaps <- gap.inspect(seqs)
mutant.sites <- mutant.sites[mutant.sites %in% gaps$f.inds]


head(mutant.sites)
```

Make a names label for our mutated resideus (one per mutation)
```{r}
mutant.names <- paste0(seqs$ali["P53_wt",mutant.sites],
                       mutant.sites,
                       seqs$ali["P53_mutant",mutant.sites])

mutant.names
```


```{r}
#sequence positions surrounding each mutant site
start.position <- mutant.sites - 8
end.position <- mutant.sites + 8


# Blank matrix to store sub-sequences
store.seqs <- matrix("-", nrow=length(mutant.sites), ncol=17)
rownames(store.seqs) <- mutant.names

# ## Extract each sub-sequence
for(i in 1:length(mutant.sites)) {
   store.seqs[i,] <- seqs$ali["P53_mutant",start.position[i]:end.position[i]]
}

mutant.sites

```

```{r}
## First blank out the gap positions 
store.seqs[store.seqs == "-"] <- ""

## Output a FASTA file for further analysis
write.fasta(seqs=store.seqs, ids=mutant.names, file="subsequences.fa")
```

