---
title: "Class 13: Genomics part I"
output: github_document
---

Load MXL SNP analysis.
```{r}
snp_data <- read.csv("MXL_snp_analysis.csv")
summary(snp_data)

```



Determine percentage of GG variants from snp_data. Use table to sort into bins and calculate percentage.
```{r}
table(snp_data$Genotype..forward.strand.) / nrow(snp_data) * 100


```

##FASTQ Quality Scores

```{r}
#install.packages("seqinr")
#install.packages("gtools")

library(seqinr)
library(gtools)
chars <- s2c("DDDDCDEDCDDDDBBDDDCC@")
chars

phred <- asc( chars ) - 33
phred
```








Load RNA seq files of 'GG' SNP patient into our environment. Let's see if this SNP changes gene expression. 
```{r}
seq1 <- read.csv("HG00109_1.fastq")
seq2 <- read.csv("HG00109_2.fastq")
#tail(seq1, col=10, rows=7)
summary(seq1)

```


##Genotype Based Expression Levels

Population scale analysis of SNP for Asthma. Load data from url with read.table(), looks better than read.csv
```{r}
expr <- read.table("https://bioboot.github.io/bggn213_W19/class-material/rs8067378_ENSG00000172057.6.txt")
head(expr)
```

```{r}
table(expr$geno)
```

```{r}
inds.gg <- expr$geno == "G/G"
summary( expr[inds.gg,"exp"] )
```

```{r}
inds.ag <- expr$geno == "A/G"
summary( expr[inds.ag,"exp"] )
```

```{r}
inds.aa <- expr$geno == "A/A"
summary( expr[inds.aa,"exp"] )

```

```{r}
boxplot(exp ~ geno, data = expr, notch = TRUE)
```

```{r}
# Boxplot with the data shown
ggplot(expr, aes(geno, exp, fill=geno)) + 
  geom_boxplot(notch=TRUE, outlier.shape = NA) + 
  geom_jitter(shape=16, position=position_jitter(0.2), alpha=0.4)
```

